RecoLocalCalo/HGCalRecAlgos/src/HGCalESProducerTools.cc

```diff
@@ -0,0 +1,95 @@
+                                      << "'";
+        return it.key();  // return matching key
+      }
+    }
+    edm::LogError("search_modkey") << "search_modkey: Could not find matching key for '" << module << "' in '" << name
+                                   << "'! Returning first key '" << data.begin().key() << "'...";
+    return data.begin().key();  // no matching key found in whole JSON map
+  }
+
+  // @short search first match to a given FED index in a JSON (following insertion order)
+  // allow glob patterns like. '1*', '1[0-5]', etc. and
+  // allow numerical ranges like '0-20', '20-40', etc.
+  std::string search_fedkey(const int& fedid, const json& data, const std::string& name = "") {
+    if (!data.is_object() or data.empty()) {
+      cms::Exception ex("InvalidData");
+      ex << "'" << name << "' does not have the expected map/dict structure!";
+      ex.addContext("Calling hgcal::search_fedkey()");
+      throw ex;
+    }
+    auto it = data.begin();
+    std::string matchedkey = data.begin().key();  // use first key as default
+    while (it != data.end()) {
+      std::string fedkey = it.key();
+
+      // try as numerical range
+      int low, high;
+      char dash;
+      std::istringstream iss(fedkey.c_str());
+      iss >> low >> dash >> high;              // parse [integer][character][integer] pattern
+      if (iss.eof() and dash == '-') {         // matches pattern
+        if (low <= fedid and fedid <= high) {  // matches numerical range
+          matchedkey = fedkey;
+          break;
+        }
+
+        // try as glob pattern
+      } else {
+        const std::string sfedid = std::to_string(fedid);
+        std::regex re(edm::glob2reg(fedkey));
+        if (std::regex_match(sfedid, re)) {  // found matching key !
+          matchedkey = fedkey;
+          break;
+        }
+      }
+
+      ++it;
+    }
+    if (it == data.end())
+      edm::LogError("search_fedkey") << "search_fedkey: Could not find matching key for '" << fedid << "' in '" << name
+                                     << "'! Returning first key '" << matchedkey << "'...";
+    else
+      edm::LogInfo("search_fedkey") << "search_fedkey: Matched module='" << fedid << "' to fedkey='" << matchedkey
+                                    << "'";
+
+    return matchedkey;  // no matching key found in whole JSON map
+  }
+
+  // @short check if JSON data contains key
+  bool check_keys(const json& data,
+                  const std::string& firstkey,
+                  const std::vector<std::string>& keys,
+                  const std::string& fname) {
+    bool iscomplete = true;
+    for (auto const& key : keys) {
+      if (not data[firstkey].contains(key)) {
+        edm::LogWarning("checkkeys") << " JSON is missing key '" << key << "' for " << firstkey << "!"
+                                     << " Please check file " << fname;
+        iscomplete = false;
+      }
+    }
+    return iscomplete;
+  }
+
+}  // namespace hgcal
```

Comment:
Can you keep a newline on the last line ?

Reply:
Done: https://github.com/cms-sw/cmssw/pull/48137/commits/dafa02849c285fcd2e4ba132baf0a21f0d1b1c59!

Reply:
It seem like `scram build code-format` prefers no newline at the end here? https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48137/44914/code-format.patch

Reply:
Mhm, `clang-format` is asking to not have an empty new line, while I think the red sign on the github diff is about a file that ends without any newline on the last line.

Reply:
As a check, could you try to
  - remove the final empty line
  - run `dos2unix` on the file
?

Or open it with `vi` and write it with no other changes ?